In [1]:
# import libraries
import pandas as pd
import keras
from sklearn.model_selection import train_test_split
from keras.models import Sequential
from keras.layers import Dense
from sklearn.metrics import mean_squared_error
import numpy as np

# load data

In [3]:
# use pandas to load data
df=pd.read_csv('concrete_data.csv')
df.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [4]:
X= df.drop('Strength',axis=1)
y=df['Strength']

In [5]:
# split data train 70% & test 30%
x_train,x_test, y_train, y_test= train_test_split(X,y,train_size=.70,random_state=42)

# Build Neural Network

In [6]:
# function of the shallow neural network

def shallow_regression ():
    #create model
    model= Sequential()
    model.add(Dense(10,activation="relu", input_shape=(x_train.shape[1],)))
    model.add(Dense(1))
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [39]:
# build the model
model=shallow_regression()

# Train Model  (A)

In [8]:
# fit the model
model.fit(x_train, y_train, epochs=50, verbose=2)

Epoch 1/50
23/23 - 1s - loss: 13609.2432 - 853ms/epoch - 37ms/step
Epoch 2/50
23/23 - 0s - loss: 8011.8628 - 36ms/epoch - 2ms/step
Epoch 3/50
23/23 - 0s - loss: 5790.4175 - 41ms/epoch - 2ms/step
Epoch 4/50
23/23 - 0s - loss: 4335.3296 - 44ms/epoch - 2ms/step
Epoch 5/50
23/23 - 0s - loss: 3024.0859 - 37ms/epoch - 2ms/step
Epoch 6/50
23/23 - 0s - loss: 1952.5372 - 37ms/epoch - 2ms/step
Epoch 7/50
23/23 - 0s - loss: 1361.3386 - 37ms/epoch - 2ms/step
Epoch 8/50
23/23 - 0s - loss: 1047.6565 - 39ms/epoch - 2ms/step
Epoch 9/50
23/23 - 0s - loss: 834.6951 - 39ms/epoch - 2ms/step
Epoch 10/50
23/23 - 0s - loss: 706.5054 - 38ms/epoch - 2ms/step
Epoch 11/50
23/23 - 0s - loss: 617.6689 - 45ms/epoch - 2ms/step
Epoch 12/50
23/23 - 0s - loss: 551.4622 - 42ms/epoch - 2ms/step
Epoch 13/50
23/23 - 0s - loss: 496.2812 - 36ms/epoch - 2ms/step
Epoch 14/50
23/23 - 0s - loss: 454.5711 - 40ms/epoch - 2ms/step
Epoch 15/50
23/23 - 0s - loss: 422.2774 - 38ms/epoch - 2ms/step
Epoch 16/50
23/23 - 0s - loss: 397.114

# Evaluate

In [9]:
loss_val = model.evaluate(x_test, y_test)
loss_val

10/10 [==============================] - 0s 2ms/step - loss: 152.4293


152.4292755126953

In [10]:
# sklearn
y_pred = model.predict(x_test)
MSE = mean_squared_error(y_test, y_pred)
mean = np.mean(MSE)
standard_deviation = np.std(MSE)
print(mean, standard_deviation)

10/10 [==============================] - 0s 3ms/step
152.42927083512157 0.0


In [11]:
n = 50
epochs = 50
mean_squared_errors = []
for i in range(0, n):
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square)

MSE 1: 129.6521453857422
10/10 [==============================] - 0s 2ms/step
MSE 2: 116.13692474365234
10/10 [==============================] - 0s 3ms/step
MSE 3: 104.88802337646484
10/10 [==============================] - 0s 2ms/step
MSE 4: 116.30975341796875
10/10 [==============================] - 0s 2ms/step
MSE 5: 122.12785339355469
10/10 [==============================] - 0s 2ms/step
MSE 6: 96.41986083984375
10/10 [==============================] - 0s 3ms/step
MSE 7: 90.90367126464844
10/10 [==============================] - 0s 2ms/step
MSE 8: 71.7380599975586
10/10 [==============================] - 0s 2ms/step
MSE 9: 66.20088958740234
10/10 [==============================] - 0s 2ms/step
MSE 10: 65.68372344970703
10/10 [==============================] - 0s 2ms/step
MSE 11: 57.58830642700195
10/10 [==============================] - 0s 2ms/step
MSE 12: 51.926124572753906
10/10 [==============================] - 0s 2ms/step
MSE 13: 59.4508171081543
10/10 [=========================

In [12]:
mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("Mean and standard deviation of " +str(n) + "\nTotal number of epochs for each training is: " +str(epochs) + "\n")
print("Mean: "+str(mean)+"\n")
print("Standard Deviation: "+str(standard_deviation))



Mean and standard deviation of 50
Total number of epochs for each training is: 50

Mean: 63.179920114198175

Standard Deviation: 21.623674234585334


# Train Model (B)

In [13]:
# normalize data
x_norm= (X-X.mean())/X.std()

In [14]:
x_norm

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0     2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1     2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2     0.491187            0.795140 -0.846733  2.174405         -1.038638   
3     0.491187            0.795140 -0.846733  2.174405         -1.038638   
4    -0.790075            0.678079 -0.846733  0.488555         -1.038638   
...        ...                 ...       ...       ...               ...   
1025 -0.045623            0.487998  0.564271 -0.092126          0.451190   
1026  0.392628           -0.856472  0.959602  0.675872          0.702285   
1027 -1.269472            0.759210  0.850222  0.521336         -0.017520   
1028 -1.168042            1.307430 -0.846733 -0.279443          0.852942   
1029 -0.193939            0.308349  0.376762  0.891286          0.400971   

      Coarse Aggregate  Fine Aggregate       Age  
0             0.862735       -1.217079 -0.279597  
1             1.055651       -1.217079 -0.279597  
2            -0.526262       -2.239829  3.551340  
3            -0.526262       -2.239829  5.055221  
4             0.070492        0.647569  4.976069  
...                ...             ...       ...  
1025         -1.322363       -0.065861 -0.279597  
1026         -1.993711        0.496651 -0.279597  
1027         -1.035561        0.080068 -0.279597  
1028          0.214537        0.191074 -0.279597  
1029         -1.394385       -0.150675 -0.279597  

[1030 rows x 8 columns]

In [15]:
# split data train 70% & test 30%
x_train,x_test, y_train, y_test= train_test_split(x_norm,y,train_size=.70,random_state=42)

In [16]:
# fit the model
model.fit(x_train, y_train, epochs=50, verbose=2)

Epoch 1/50
23/23 - 0s - loss: 1571.5677 - 80ms/epoch - 3ms/step
Epoch 2/50
23/23 - 0s - loss: 1537.8937 - 57ms/epoch - 2ms/step
Epoch 3/50
23/23 - 0s - loss: 1500.2507 - 65ms/epoch - 3ms/step
Epoch 4/50
23/23 - 0s - loss: 1459.9484 - 49ms/epoch - 2ms/step
Epoch 5/50
23/23 - 0s - loss: 1417.4471 - 52ms/epoch - 2ms/step
Epoch 6/50
23/23 - 0s - loss: 1373.5847 - 36ms/epoch - 2ms/step
Epoch 7/50
23/23 - 0s - loss: 1330.6044 - 39ms/epoch - 2ms/step
Epoch 8/50
23/23 - 0s - loss: 1286.7402 - 41ms/epoch - 2ms/step
Epoch 9/50
23/23 - 0s - loss: 1243.8713 - 104ms/epoch - 5ms/step
Epoch 10/50
23/23 - 0s - loss: 1201.5140 - 53ms/epoch - 2ms/step
Epoch 11/50
23/23 - 0s - loss: 1160.4434 - 64ms/epoch - 3ms/step
Epoch 12/50
23/23 - 0s - loss: 1120.8271 - 58ms/epoch - 3ms/step
Epoch 13/50
23/23 - 0s - loss: 1082.5724 - 60ms/epoch - 3ms/step
Epoch 14/50
23/23 - 0s - loss: 1044.6257 - 64ms/epoch - 3ms/step
Epoch 15/50
23/23 - 0s - loss: 1008.5641 - 61ms/epoch - 3ms/step
Epoch 16/50
23/23 - 0s - loss: 97

# Evaluate

In [17]:
loss_val = model.evaluate(x_test, y_test)
loss_val

10/10 [==============================] - 0s 3ms/step - loss: 311.7969


311.79693603515625

In [18]:
# sklearn
y_pred = model.predict(x_test)
MSE = mean_squared_error(y_test, y_pred)
mean = np.mean(MSE)
standard_deviation = np.std(MSE)
print(mean, standard_deviation)

10/10 [==============================] - 0s 2ms/step
311.79691169672714 0.0


In [19]:
n = 50
epochs = 50
mean_squared_errors = []
for i in range(0, n):
    X_train, X_test, y_train, y_test = train_test_split(x_norm,y, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square)

MSE 1: 147.6158447265625
10/10 [==============================] - 0s 2ms/step
MSE 2: 151.8385467529297
10/10 [==============================] - 0s 2ms/step
MSE 3: 123.94319915771484
10/10 [==============================] - 0s 2ms/step
MSE 4: 125.19348907470703
10/10 [==============================] - 0s 3ms/step
MSE 5: 116.38221740722656
10/10 [==============================] - 0s 2ms/step
MSE 6: 91.23208618164062
10/10 [==============================] - 0s 2ms/step
MSE 7: 100.837646484375
10/10 [==============================] - 0s 2ms/step
MSE 8: 66.23623657226562
10/10 [==============================] - 0s 2ms/step
MSE 9: 66.95246887207031
10/10 [==============================] - 0s 2ms/step
MSE 10: 56.85020446777344
10/10 [==============================] - 0s 2ms/step
MSE 11: 48.08860778808594
10/10 [==============================] - 0s 2ms/step
MSE 12: 46.650115966796875
10/10 [==============================] - 0s 2ms/step
MSE 13: 52.59178161621094
10/10 [=========================

In [20]:
mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("Mean and standard deviation of " +str(n) + "\nTotal number of epochs for each training is: " +str(epochs) + "\n")
print("Mean: "+str(mean)+"\n")
print("Standard Deviation: "+str(standard_deviation))



Mean and standard deviation of 50
Total number of epochs for each training is: 50

Mean: 53.18035674001306

Standard Deviation: 29.83054824653526


# Train Model (C)

In [21]:
# normalize data
x_norm= (X-X.mean())/X.std()

In [22]:
x_norm

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0     2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1     2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2     0.491187            0.795140 -0.846733  2.174405         -1.038638   
3     0.491187            0.795140 -0.846733  2.174405         -1.038638   
4    -0.790075            0.678079 -0.846733  0.488555         -1.038638   
...        ...                 ...       ...       ...               ...   
1025 -0.045623            0.487998  0.564271 -0.092126          0.451190   
1026  0.392628           -0.856472  0.959602  0.675872          0.702285   
1027 -1.269472            0.759210  0.850222  0.521336         -0.017520   
1028 -1.168042            1.307430 -0.846733 -0.279443          0.852942   
1029 -0.193939            0.308349  0.376762  0.891286          0.400971   

      Coarse Aggregate  Fine Aggregate       Age  
0             0.862735       -1.217079 -0.279597  
1             1.055651       -1.217079 -0.279597  
2            -0.526262       -2.239829  3.551340  
3            -0.526262       -2.239829  5.055221  
4             0.070492        0.647569  4.976069  
...                ...             ...       ...  
1025         -1.322363       -0.065861 -0.279597  
1026         -1.993711        0.496651 -0.279597  
1027         -1.035561        0.080068 -0.279597  
1028          0.214537        0.191074 -0.279597  
1029         -1.394385       -0.150675 -0.279597  

[1030 rows x 8 columns]

In [23]:
# split data train 70% & test 30%
x_train,x_test, y_train, y_test= train_test_split(x_norm,y,train_size=.70,random_state=42)

In [24]:
# fit the model
model.fit(x_train, y_train, epochs=100, verbose=2)

Epoch 1/100
23/23 - 0s - loss: 35.0347 - 65ms/epoch - 3ms/step
Epoch 2/100
23/23 - 0s - loss: 34.8293 - 66ms/epoch - 3ms/step
Epoch 3/100
23/23 - 0s - loss: 34.8418 - 68ms/epoch - 3ms/step
Epoch 4/100
23/23 - 0s - loss: 34.7033 - 48ms/epoch - 2ms/step
Epoch 5/100
23/23 - 0s - loss: 34.7352 - 40ms/epoch - 2ms/step
Epoch 6/100
23/23 - 0s - loss: 34.7128 - 49ms/epoch - 2ms/step
Epoch 7/100
23/23 - 0s - loss: 34.6473 - 42ms/epoch - 2ms/step
Epoch 8/100
23/23 - 0s - loss: 34.6684 - 49ms/epoch - 2ms/step
Epoch 9/100
23/23 - 0s - loss: 34.6854 - 38ms/epoch - 2ms/step
Epoch 10/100
23/23 - 0s - loss: 34.6506 - 51ms/epoch - 2ms/step
Epoch 11/100
23/23 - 0s - loss: 34.6405 - 38ms/epoch - 2ms/step
Epoch 12/100
23/23 - 0s - loss: 34.7125 - 49ms/epoch - 2ms/step
Epoch 13/100
23/23 - 0s - loss: 34.6275 - 38ms/epoch - 2ms/step
Epoch 14/100
23/23 - 0s - loss: 34.7244 - 56ms/epoch - 2ms/step
Epoch 15/100
23/23 - 0s - loss: 34.6057 - 37ms/epoch - 2ms/step
Epoch 16/100
23/23 - 0s - loss: 34.6037 - 39ms/ep

# Evaluate

In [25]:
loss_val = model.evaluate(x_test, y_test)
loss_val

10/10 [==============================] - 0s 3ms/step - loss: 37.0757


37.075740814208984

In [26]:
# sklearn
y_pred = model.predict(x_test)
MSE = mean_squared_error(y_test, y_pred)
mean = np.mean(MSE)
standard_deviation = np.std(MSE)
print(mean, standard_deviation)

10/10 [==============================] - 0s 2ms/step
37.075743617179775 0.0


In [40]:
n = 50
epochs = 100
mean_squared_errors = []
for i in range(0, n):
    X_train, X_test, y_train, y_test = train_test_split(x_norm,y, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square)

MSE 1: 147.9617919921875
10/10 [==============================] - 0s 3ms/step
MSE 2: 107.93855285644531
10/10 [==============================] - 0s 2ms/step
MSE 3: 67.70933532714844
10/10 [==============================] - 0s 2ms/step
MSE 4: 53.97392654418945
10/10 [==============================] - 0s 2ms/step
MSE 5: 48.37443542480469
10/10 [==============================] - 0s 3ms/step
MSE 6: 51.56919860839844
10/10 [==============================] - 0s 8ms/step
MSE 7: 53.549556732177734
10/10 [==============================] - 0s 2ms/step
MSE 8: 41.333980560302734
10/10 [==============================] - 0s 2ms/step
MSE 9: 43.901527404785156
10/10 [==============================] - 0s 2ms/step
MSE 10: 43.54914474487305
10/10 [==============================] - 0s 3ms/step
MSE 11: 40.533145904541016
10/10 [==============================] - 0s 2ms/step
MSE 12: 38.032745361328125
10/10 [==============================] - 0s 3ms/step
MSE 13: 46.830284118652344
10/10 [=====================

In [41]:
mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("Mean and standard deviation of " +str(n) + "\nTotal number of epochs for each training is: " +str(epochs) + "\n")
print("Mean: "+str(mean)+"\n")
print("Standard Deviation: "+str(standard_deviation))



Mean and standard deviation of 50
Total number of epochs for each training is: 100

Mean: 45.5496644167582

Standard Deviation: 18.207118280702815


#Build  Deep Neural Network

In [29]:
# function of the Deep neural network

def Deep_regression ():
    #create model
    model= Sequential()
    model.add(Dense(10,activation="relu", input_shape=(x_train.shape[1],)))
    model.add(Dense(10,activation="relu"))
    model.add(Dense(10,activation="relu"))

    model.add(Dense(1))
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

# Train Model (D)

In [30]:
# normalize data
x_norm= (X-X.mean())/X.std()

In [31]:
x_norm

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0     2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1     2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2     0.491187            0.795140 -0.846733  2.174405         -1.038638   
3     0.491187            0.795140 -0.846733  2.174405         -1.038638   
4    -0.790075            0.678079 -0.846733  0.488555         -1.038638   
...        ...                 ...       ...       ...               ...   
1025 -0.045623            0.487998  0.564271 -0.092126          0.451190   
1026  0.392628           -0.856472  0.959602  0.675872          0.702285   
1027 -1.269472            0.759210  0.850222  0.521336         -0.017520   
1028 -1.168042            1.307430 -0.846733 -0.279443          0.852942   
1029 -0.193939            0.308349  0.376762  0.891286          0.400971   

      Coarse Aggregate  Fine Aggregate       Age  
0             0.862735       -1.217079 -0.279597  
1             1.055651       -1.217079 -0.279597  
2            -0.526262       -2.239829  3.551340  
3            -0.526262       -2.239829  5.055221  
4             0.070492        0.647569  4.976069  
...                ...             ...       ...  
1025         -1.322363       -0.065861 -0.279597  
1026         -1.993711        0.496651 -0.279597  
1027         -1.035561        0.080068 -0.279597  
1028          0.214537        0.191074 -0.279597  
1029         -1.394385       -0.150675 -0.279597  

[1030 rows x 8 columns]

In [32]:
# split data train 70% & test 30%
x_train,x_test, y_train, y_test= train_test_split(x_norm,y,train_size=.70,random_state=42)

In [33]:
model = Deep_regression()

In [34]:
# fit the model
model.fit(x_train, y_train, epochs=50, verbose=2)

Epoch 1/50
23/23 - 1s - loss: 1595.8710 - 961ms/epoch - 42ms/step
Epoch 2/50
23/23 - 0s - loss: 1574.5686 - 49ms/epoch - 2ms/step
Epoch 3/50
23/23 - 0s - loss: 1547.8094 - 48ms/epoch - 2ms/step
Epoch 4/50
23/23 - 0s - loss: 1506.9556 - 54ms/epoch - 2ms/step
Epoch 5/50
23/23 - 0s - loss: 1439.3889 - 45ms/epoch - 2ms/step
Epoch 6/50
23/23 - 0s - loss: 1328.4098 - 50ms/epoch - 2ms/step
Epoch 7/50
23/23 - 0s - loss: 1159.4437 - 45ms/epoch - 2ms/step
Epoch 8/50
23/23 - 0s - loss: 936.4466 - 48ms/epoch - 2ms/step
Epoch 9/50
23/23 - 0s - loss: 692.1418 - 49ms/epoch - 2ms/step
Epoch 10/50
23/23 - 0s - loss: 489.7902 - 44ms/epoch - 2ms/step
Epoch 11/50
23/23 - 0s - loss: 371.5894 - 45ms/epoch - 2ms/step
Epoch 12/50
23/23 - 0s - loss: 313.9086 - 38ms/epoch - 2ms/step
Epoch 13/50
23/23 - 0s - loss: 275.4717 - 46ms/epoch - 2ms/step
Epoch 14/50
23/23 - 0s - loss: 247.3333 - 40ms/epoch - 2ms/step
Epoch 15/50
23/23 - 0s - loss: 227.6955 - 42ms/epoch - 2ms/step
Epoch 16/50
23/23 - 0s - loss: 213.2856 

# Evaluate

In [35]:
loss_val = model.evaluate(x_test, y_test)
loss_val

10/10 [==============================] - 0s 3ms/step - loss: 135.3576


135.3575897216797

In [36]:
# sklearn
y_pred = model.predict(x_test)
MSE = mean_squared_error(y_test, y_pred)
mean = np.mean(MSE)
standard_deviation = np.std(MSE)
print(mean, standard_deviation)

10/10 [==============================] - 0s 3ms/step
135.35758822183686 0.0


In [37]:
n = 50
epochs = 50
mean_squared_errors = []
for i in range(0, n):
    X_train, X_test, y_train, y_test = train_test_split(x_norm,y, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square)

MSE 1: 104.17884063720703
10/10 [==============================] - 0s 2ms/step
MSE 2: 118.37779235839844
10/10 [==============================] - 0s 3ms/step
MSE 3: 101.85169219970703
10/10 [==============================] - 0s 2ms/step
MSE 4: 88.5061264038086
10/10 [==============================] - 0s 3ms/step
MSE 5: 60.76948547363281
10/10 [==============================] - 0s 2ms/step
MSE 6: 53.85230255126953
10/10 [==============================] - 0s 3ms/step
MSE 7: 50.75822830200195
10/10 [==============================] - 0s 2ms/step
MSE 8: 36.26168441772461
10/10 [==============================] - 0s 2ms/step
MSE 9: 35.883323669433594
10/10 [==============================] - 0s 2ms/step
MSE 10: 36.87004089355469
10/10 [==============================] - 0s 2ms/step
MSE 11: 32.77991485595703
10/10 [==============================] - 0s 2ms/step
MSE 12: 26.714536666870117
10/10 [==============================] - 0s 3ms/step
MSE 13: 35.163352966308594
10/10 [=======================

In [38]:
mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("Mean and standard deviation of " +str(n) + "\nTotal number of epochs for each training is: " +str(epochs) + "\n")
print("Mean: "+str(mean)+"\n")
print("Standard Deviation: "+str(standard_deviation))



Mean and standard deviation of 50
Total number of epochs for each training is: 50

Mean: 35.2331185674546

Standard Deviation: 21.683061920557957


***************************************************************


Task A

    Mean = 63.179920114198175

Task B
    
    Mean= 53.18035674001306

Task C
    
    Mean= 45.5496644167582

Task D
    
    Mean= 35.2331185674546